cmd/virt-prerunner/main.go

```diff
@@ -101,6 +110,24 @@ func buildVMConfig(ctx context.Context, vm *virtv1alpha1.VirtualMachine) (*cloud
 		}
 	}
 
+	if vm.Spec.Instance.CPU.DedicatedCPUPlacement {
+		cpuSet, err := cpuset.Get()
+		if err != nil {
+			return nil, fmt.Errorf("get CPU set: %s", err)
+		}
+
+		pcpus := cpuSet.ToSlice()
+		numPCPUs := len(pcpus)
+		numVCPUs := int(vm.Spec.Instance.CPU.Sockets * vm.Spec.Instance.CPU.CoresPerSocket)
+
+		for i := 0; i < numVCPUs; i++ {
+			vmConfig.Cpus.Affinity = append(vmConfig.Cpus.Affinity, &cloudhypervisor.CpuAffinity{
+				Vcpu:     i,
+				HostCpus: []int{pcpus[i%numPCPUs]},
```

Comment:
Should we add a webhook validation for CPU resources? If the CPU resource is not guaranteed, it may not be a good idea to bind vCPUs here?

Reply:
It's optional for user to configure kubelet CPU manager policy to `Static` and use `Guaranteed` pod to get exclusive CPUs on the node, but user can get better performance by using `Static` policy and `Guaranteed` QOS. Without any configuration, we bind vCPU to avoid cache missing causing by CPU migration.

Reply:
If the VM pod is not in the Guaranteed group, it could see all PCPUs are available for binding? If so, the `i%numPCPUs` part would always choose to bind the first several PCPUs?

Reply:
Right, maybe a random algorithm of binding vCPUs could be better. For example, binding vCPUs[0,1,2,3,4,5] to pCPUs[0,1,2,3] can be resolved as follows:

step1: Use each pCPU in a balanced manner. vCPU -> pCPU [0 -> 0, 1 -> 1, 2 -> 2, 3 -> 3]
step2: Bind the other vCPUs randomly. vCPU -> pCPU [4 -> one of 0~3, 5 -> one of 0~3]

Reply:
I'm still having a bad feeling about this. It's neither dedicated nor efficient to bind vCPUs to physical cores if they have unmatched numbers.